cmd/tui/selectable_table.go

```diff
@@ -0,0 +1,475 @@
+// SendUpdate sends a synchronous update to the table. It waits until the program exists.
+func (s *selectableTable) SendUpdate(msg tea.Msg) {
+	// Sleep until the program is set.
+	for s.program == nil {
+		time.Sleep(300 * time.Millisecond)
+	}
+
+	s.program.Send(msg)
+}
+
+// Update handles table updates.
+func (s *selectableTable) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	switch msg := msg.(type) {
+	case tea.KeyMsg:
+		return s.handleKeyEvent(msg)
+	case InsertMsg:
+		return s.handleInsertEvent(msg)
+	case RemoveMsg:
+		return s.handleRemoveEvent(msg)
+	case EnableMsg:
+		return s.handleEnableEvent(msg)
+	case DisableMsg:
+		return s.handleDisableEvent(msg)
+	case ErrorMsg:
+		s.err = msg
+
+		return s, tea.Quit
+	}
+
+	return s, nil
+}
+
+// Init initializes a new selectable table.
+func (s *selectableTable) Init() tea.Cmd {
+	s.activeRows = make(map[int]bool)
+	s.disabledRows = make(map[int]bool)
+	s.formatRows = make([][]string, 0, len(s.rawRows))
+	s.filterMap = make(map[int]int, len(s.rawRows))
+
+	s.currentRow = 0
+	s.startIndex = 0
+	s.filter = ""
+
+	for i, row := range s.rawRows {
+		s.filterMap[i] = i
+		col := make([]string, 0, len(row))
+		col = append(col, row...)
+
+		s.formatRows = append(s.formatRows, col)
+	}
+
+	header := make([]string, len(s.header))
+	copy(header, s.header)
+
+	s.table = baseTableTemplate(header)
+	s.updateTableRows()
+	return nil
+}
+
+// View draws the table and its menus and returns it as a string.
+func (s *selectableTable) View() string {
+	if !s.active {
+		return ""
+	}
+
+	tableStr := s.table.String()
+	parts := strings.Split(tableStr, "\n")
+
+	// These are the number of rows taken up by the table header and footer.
+	headerLength := 3
+	footerLength := 1
+
+	for i, part := range parts {
+		row := i + s.startIndex
+		var selector string
+		if i == 2 && s.startIndex > 0 {
+			selector = lipgloss.NewStyle().SetString("↑").Bold(true).Foreground(White).String()
+		} else if i == len(parts)-footerLength && s.startIndex < len(s.formatRows)-s.size {
+			selector = lipgloss.NewStyle().SetString("↓").Bold(true).Foreground(White).String()
+		} else if i == len(parts)-footerLength {
+			selector = " "
+		} else if i < headerLength {
+			selector = " "
+		} else if s.activeRows[s.filterMap[row-headerLength]] {
+			selector = SuccessSymbol()
+		} else if s.disabledRows[s.filterMap[row-headerLength]] {
+			selector = ErrorSymbol()
+		} else if s.currentRow+headerLength == row {
+			selector = SetColor(Yellow, ">", false)
+		} else {
+			selector = " "
+		}
+
+		parts[i] = fmt.Sprintf(" %s %s", selector, part)
+	}
+
+	title := s.title
+	if title != "" {
+		title = lipgloss.NewStyle().SetString(s.title).Bold(true).String()
+	}
+
+	// If the test console is active, just print the title so that the table output isn't corrupted by debug messages.
+	if os.Getenv("TEST_CONSOLE") == "1" {
+		return title + "\n"
+	}
+
+	filter := Printf(Fmt{Arg: "\n Filter | %s\n", Color: White}, Fmt{Arg: s.filter, Color: Yellow})
```

Comment:
Not following, if you mean just mention that you can type, that's already at the bottom of the table. If you mean add a blinking cursor at the end, I'll give it a try but it might get a bit complex with terminal escape codes.

Reply:
I've moved the hint from below the table to be in-line with the filter prompt.